### ISSUE TITLE
Disk swap interface is causing crashes

### ISSUE THREAD

**Issue by jet082** [libretro/quasi88-libretro#9]
For games like Ys, where you have to swap from Disk A to Disk B, it is impossible to complete on RetroArch currently.

**Comment by celerizer** [comment]
Hello, try it now using the subsystem interface. When you load 3 disks you can now use L/R and the d-pad to cycle between them.

**Comment by jet082** [comment]
Unfortunately, when I press l1 or r1, the game crashes. Is there something else that I need to do?

**Comment by celerizer** [comment]
When you say the game crashes do you mean the core itself or the emulated game? if you could provide more info that'd be very useful

**Comment by jet082** [comment]
I mean the core/retroarch crashes. Here is the file I am testing against. I also ran the core through a debugger and will include the relevant debug logs:

[ys1.zip](https://github.com/libretro/quasi88-libretro/files/3692836/ys1.zip)

https://hastebin.com/rejakujane.xml
https://hastebin.com/huporebire.xml
https://hastebin.com/jopuyifato.xml
https://hastebin.com/oluyutapav.xml (this is likely the most relevant)

Error: Thread 1: EXC_BREAKPOINT (code=1, subcode=0x18d2f92b0)

**Comment by Sanaki** [comment]
Experiencing the same thing myself. Happy to provide whatever may be needed to diagnose.

**Comment by celerizer** [comment]
What platform is this on? I've tried on everything I own and can't seem to get it to crash like this

**Comment by jet082** [comment]
To clarify - have you tested this on multi-disk .d88 files? The file I have been using has Disk A, Disk B, and the User Disk all bundled into a single .d88.

**Comment by Sanaki** [comment]
Mint 19.2, kernel 4.15.0-66-generic. I don't have any multidisk single files and am not sure how to concatenate them. I've been either loading multiple disks via the menu options or via m3u.

**Comment by Sanaki** [comment]
[ra.txt](https://github.com/libretro/quasi88-libretro/files/3785313/ra.txt)
In case it's relevant, a log of starting retroarch and triggering the crash.

**Comment by jet082** [comment]
My crash is occurring on iOS.

**Comment by celerizer** [comment]
I didn't know anything about concatenated D88 files but as far as I know they shouldn't work and aren't related to this issue. Must just be messy string handling for the filenames

**Comment by Sanaki** [comment]
Just in case I'm interpreting your comment correctly, I tried renaming Hard Rank's files to HRA.d88 and HRB.d88, same behavior persists.

**Comment by a5723797** [comment]
Same on Nintendo Switch version. Subsystem does not work at all. After choosing "start N-disk game", the screen turned black and return to the Retroarch menu  immediately. It never started loading disks.

BTW, I tried same multi-disk games on windows version, it works perfectly.

**Comment by celerizer** [comment]
@a5723797

This is a problem with RetroArch, not this core. M3U playlists need to be used on consoles that statically compile cores into RA. See https://github.com/libretro/quasi88-libretro/issues/3

**Comment by gbaway** [comment]
I ran into a similar issue while running this on Linux and posted a quick fix here: [Swapping disks causes a buffer overflow](https://github.com/libretro/quasi88-libretro/issues/49)

**Comment by Sanaki** [comment]
That doesn't look "similar", it looks like that is this issue.

**Comment by gbaway** [comment]
> That doesn't look "similar", it looks like that is this issue. 

Well, a more appropiate way to solve this crash is to replace the variable msg's type from char to wchar.
This is because, while inspecting the string in gdb, there where a lot of UTF-8 garbage characters.

Can you try applying this patch to **src/LIBRETRO/disks.c** to see if this solves the crash for you aswell?

```
4a5
> #include <wchar.h>
24c25
<       char    msg[OSD_MAX_FILENAME];
---
>       wchar_t msg[OSD_MAX_FILENAME];
```

Cheers.

[overflow.patch.txt](https://github.com/libretro/quasi88-libretro/files/7028133/overflow.patch.txt)

**Comment by Sanaki** [comment]
That does indeed fix it. Please submit a PR.

**Comment by gbaway** [comment]
@Sanaki done: [Update disks.c - Fixes disk swap interface causing crashes #50](https://github.com/libretro/quasi88-libretro/pull/50)

**Comment by Sanaki** [comment]
The above PR was merged, so this issue can now be closed.
